pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -1921,7 +1934,59 @@ impl RemoteTimelineClient {
             }
 
             let upload_result: anyhow::Result<()> = match &task.op {
-                UploadOp::UploadLayer(ref layer, ref layer_metadata) => {
+                UploadOp::UploadLayer(ref layer, ref layer_metadata, last_op) => {
+                    if let Some(OpType::FlushDeletion) = last_op {
+                        if self.config.read().unwrap().block_deletions {
+                            // Of course, this is not efficient... but usually the queue should be empty.
+                            let mut queue_locked = self.upload_queue.lock().unwrap();
+                            let mut detected = false;
+                            if let Ok(queue) = queue_locked.initialized_mut() {
+                                for list in queue.blocked_deletions.iter_mut() {
+                                    list.layers.retain(|(name, meta)| {
+                                        if name == &layer.layer_desc().layer_name()
+                                            && meta.generation == layer_metadata.generation
+                                        {
+                                            detected = true;
+                                            // remove the layer from deletion queue
+                                            false
+                                        } else {
+                                            // keep the layer
+                                            true
+                                        }
+                                    });
+                                }
+                            }
+                            if detected {
+                                info!(
+                                    "upload_queue_reordering: cancelled blocked deletion of layer {} at gen {:?}",
```

Comment:
`upload_queue_reordering: ` is a bit abstract (we're not reordering anything here, just reasoning about whether it is safe to do so).  I would either drop the prefix, or make it "perform_upload_task:" just to make it obvious where the callsite is.

Reply:
https://github.com/neondatabase/neon/pull/9844/commits/2a7d00296edc3c7adea82227080c8bbeadb7688e